File: sdk/java/src/main/java/io/ferritelabs/ferrite/FerriteClient.java
package io.ferritelabs.ferrite;

import io.ferritelabs.ferrite.pool.ConnectionPool;

import java.io.Closeable;
import java.io.IOException;
import java.time.Duration;
import java.util.*;
import java.util.stream.Collectors;

/**
 * Main Ferrite client with connection pooling and support for all standard
 * Redis data structures plus Ferrite-specific extensions.
 *
 * <pre>{@code
 * FerriteConfig config = FerriteConfig.builder()
 *     .host("localhost")
 *     .port(6379)
 *     .build();
 * try (FerriteClient client = FerriteClient.connect(config)) {
 *     client.set("key", "value");
 *     String val = client.get("key");
 * }
 * }</pre>
 */
public class FerriteClient implements Closeable {

    private final ConnectionPool pool;
    private final FerriteConfig config;

    private FerriteClient(FerriteConfig config) {
        this.config = config;
        this.pool = new ConnectionPool(config);
    }

    /**
     * Connects to a Ferrite server with the given configuration.
     */
    public static FerriteClient connect(FerriteConfig config) throws IOException {
        FerriteClient client = new FerriteClient(config);
        // Validate connectivity
        client.ping();
        // Authenticate if needed
        if (config.getPassword() != null && !config.getPassword().isEmpty()) {
            client.execute("AUTH", config.getPassword());
        }
        // Select database
        if (config.getDatabase() != 0) {
            client.execute("SELECT", String.valueOf(config.getDatabase()));
        }
        return client;
    }

    /**
     * Connects to a Ferrite server at the given host and port with defaults.
     */
    public static FerriteClient connect(String host, int port) throws IOException {
        return connect(FerriteConfig.builder().host(host).port(port).build());
    }

    @Override
    public void close() throws IOException {
        pool.close();
    }

    // --- Raw command execution ---

    /**
     * Executes a raw RESP command. Used internally and by extension clients
     * (VectorClient, QueryClient).
     */
    Object executeRaw(String... args) throws IOException {
        return execute(args);
    }

    private Object execute(String... args) throws IOException {
        IOException lastErr = null;
        for (int attempt = 0; attempt <= config.getMaxRetries(); attempt++) {
            if (attempt > 0) {
                try {
                    Thread.sleep(config.getRetryBackoff().toMillis() * attempt);
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    throw new IOException("Interrupted during retry", e);
                }
            }
            ConnectionPool.Connection conn = pool.get();
            try {
                conn.getEncoder().writeCommand(args);
                Object result = conn.getDecoder().readValue();
                pool.release(conn);
                return result;
            } catch (IOException e) {
                lastErr = e;
                try { conn.close(); } catch (IOException ignored) {}
            }
        }
        throw lastErr;
    }

    // --- String operations ---

    /** Gets the value of a key. Returns null if the key does not exist. */
    public String get(String key) throws IOException {
        Object val = execute("GET", key);
        return val != null ? val.toString() : null;
    }

    /** Sets a key to a string value. */
    public void set(String key, String value) throws IOException {
        execute("SET", key, value);
    }

    /** Sets a key to a string value with a TTL. */
    public void set(String key, String value, Duration ttl) throws IOException {
        execute("SET", key, value, "PX", String.valueOf(ttl.toMillis()));
    }

    /** Deletes one or more keys. Returns the number of keys removed. */
    public long del(String... keys) throws IOException {
        String[] args = new String[keys.length + 1];
        args[0] = "DEL";
        System.arraycopy(keys, 0, args, 1, keys.length);
        return (Long) execute(args);
    }

    /** Returns the number of specified keys that exist. */
    public long exists(String... keys) throws IOException {
        String[] args = new String[keys.length + 1];
        args[0] = "EXISTS";
        System.arraycopy(keys, 0, args, 1, keys.length);
        return (Long) execute(args);
    }

    /** Sets a timeout on a key in seconds. Returns true if the timeout was set. */
    public boolean expire(String key, Duration ttl) throws IOException {
        return (Long) execute("EXPIRE", key, String.valueOf(ttl.getSeconds())) == 1;
    }

    /** Returns the remaining TTL of a key in seconds. */
    public long ttl(String key) throws IOException {
        return (Long) execute("TTL", key);
    }

    // --- List operations ---

    /** Prepends values to a list. Returns the new list length. */
    public long lpush(String key, String... values) throws IOException {
        String[] args = new String[values.length + 2];
        args[0] = "LPUSH";
        args[1] = key;
        System.arraycopy(values, 0, args, 2, values.length);
        return (Long) execute(args);
    }

    /** Appends values to a list. Returns the new list length. */
    public long rpush(String key, String... values) throws IOException {
        String[] args = new String[values.length + 2];
        args[0] = "RPUSH";
        args[1] = key;
        System.arraycopy(values, 0, args, 2, values.length);
        return (Long) execute(args);
    }

    /** Removes and returns the first element. Returns null if list is empty. */
    public String lpop(String key) throws IOException {
        Object val = execute("LPOP", key);
        return val != null ? val.toString() : null;
    }

    /** Removes and returns the last element. Returns null if list is empty. */
    public String rpop(String key) throws IOException {
        Object val = execute("RPOP", key);
        return val != null ? val.toString() : null;
    }

    /** Returns elements in the given range. */
    @SuppressWarnings("unchecked")
    public List<String> lrange(String key, long start, long stop) throws IOException {
        Object val = execute("LRANGE", key, String.valueOf(start), String.valueOf(stop));
        return toStringList(val);
    }

    /** Returns the length of a list. */
    public long llen(String key) throws IOException {
        return (Long) execute("LLEN", key);
    }

    // --- Hash operations ---

    /** Sets field-value pairs in a hash. Returns the number of new fields added. */
    public long hset(String key, String... fieldValues) throws IOException {
        String[] args = new String[fieldValues.length + 2];
        args[0] = "HSET";
        args[1] = key;
        System.arraycopy(fieldValues, 0, args, 2, fieldValues.length);
        return (Long) execute(args);
    }

    /** Gets the value of a hash field. Returns null if field does not exist. */
    public String hget(String key, String field) throws IOException {
        Object val = execute("HGET", key, field);
        return val != null ? val.toString() : null;
    }

    /** Returns all fields and values of a hash. */
    @SuppressWarnings("unchecked")
    public Map<String, String> hgetAll(String key) throws IOException {
        Object val = execute("HGETALL", key);
        return toStringMap(val);
    }

    /** Deletes fields from a hash. Returns the number removed. */
    public long hdel(String key, String... fields) throws IOException {
        String[] args = new String[fields.length + 2];
        args[0] = "HDEL";
        args[1] = key;
        System.arraycopy(fields, 0, args, 2, fields.length);
        return (Long) execute(args);
    }

    // --- Set operations ---

    /** Adds members to a set. Returns the number of new members added. */
    public long sadd(String key, String... members) throws IOException {
        String[] args = new String[members.length + 2];
        args[0] = "SADD";
        args[1] = key;
        System.arraycopy(members, 0, args, 2, members.length);
        return (Long) execute(args);
    }

    /** Returns all members of a set. */
    @SuppressWarnings("unchecked")
    public Set<String> smembers(String key) throws IOException {
        Object val = execute("SMEMBERS", key);
        List<String> list = toStringList(val);
        return new HashSet<>(list);
    }

    /** Removes members from a set. Returns the number removed. */
    public long srem(String key, String... members) throws IOException {
        String[] args = new String[members.length + 2];
        args[0] = "SREM";
        args[1] = key;
        System.arraycopy(members, 0, args, 2, members.length);
        return (Long) execute(args);
    }

    /** Returns the number of members in a set. */
    public long scard(String key) throws IOException {
        return (Long) execute("SCARD", key);
    }

    // --- Sorted set operations ---

    /** Adds a member with a score. Returns the number of new members added. */
    public long zadd(String key, double score, String member) throws IOException {
        return (Long) execute("ZADD", key, String.valueOf(score), member);
    }

    /** Returns members in the given rank range. */
    @SuppressWarnings("unchecked")
    public List<String> zrange(String key, long start, long stop) throws IOException {
        Object val = execute("ZRANGE", key, String.valueOf(start), String.valueOf(stop));
        return toStringList(val);
    }

    /** Returns the score of a member. Returns null if not found. */
    public Double zscore(String key, String member) throws IOException {
        Object val = execute("ZSCORE", key, member);
        if (val == null) return null;
        return Double.parseDouble(val.toString());
    }

    /** Removes members from a sorted set. Returns the number removed. */
    public long zrem(String key, String... members) throws IOException {
        String[] args = new String[members.length + 2];
        args[0] = "ZREM";
        args[1] = key;
        System.arraycopy(members, 0, args, 2, members.length);
        return (Long) execute(args);
    }

    // --- Server operations ---

    /** Sends PING to verify connectivity. */
    public void ping() throws IOException {
        execute("PING");
    }

    /** Returns server information. */
    public String info() throws IOException {
        Object val = execute("INFO");
        return val != null ? val.toString() : "";
    }

    // --- Helpers ---

    @SuppressWarnings("unchecked")
    private static List<String> toStringList(Object val) {
        if (val == null) return Collections.emptyList();
        List<Object> list = (List<Object>) val;
        return list.stream()
                .map(o -> o != null ? o.toString() : null)
                .collect(Collectors.toList());
    }

    @SuppressWarnings("unchecked")
    private static Map<String, String> toStringMap(Object val) {
        if (val == null) return Collections.emptyMap();
        List<Object> list = (List<Object>) val;
        Map<String, String> map = new LinkedHashMap<>();
        for (int i = 0; i < list.size() - 1; i += 2) {
            String k = list.get(i) != null ? list.get(i).toString() : null;
            String v = list.get(i + 1) != null ? list.get(i + 1).toString() : null;
            map.put(k, v);
        }
        return map;
    }
}

File: sdk/dotnet/Ferrite.Client/FerriteClient.cs
using System.Globalization;
using Ferrite.Client.Protocol;

namespace Ferrite.Client;

/// <summary>
/// Main Ferrite client with async operations and connection pooling.
/// Implements <see cref="IAsyncDisposable"/> for proper resource cleanup.
/// </summary>
/// <example>
/// <code>
/// await using var client = await FerriteClient.ConnectAsync("localhost", 6379);
/// await client.SetAsync("key", "value");
/// var val = await client.GetAsync("key");
/// </code>
/// </example>
public sealed class FerriteClient : IAsyncDisposable
{
    private readonly ConnectionPool _pool;
    private readonly FerriteOptions _options;

    private FerriteClient(FerriteOptions options)
    {
        _options = options;
        _pool = new ConnectionPool(options);
    }

    /// <summary>
    /// Connects to a Ferrite server with the specified options.
    /// </summary>
    public static async Task<FerriteClient> ConnectAsync(
        string host, int port, FerriteOptions? options = null, CancellationToken ct = default)
    {
        options ??= new FerriteOptions();
        options.Host = host;
        options.Port = port;

        var client = new FerriteClient(options);

        // Validate connectivity
        await client.PingAsync(ct).ConfigureAwait(false);

        // Authenticate
        if (!string.IsNullOrEmpty(options.Password))
        {
            await client.ExecuteAsync(ct, "AUTH", options.Password).ConfigureAwait(false);
        }

        // Select database
        if (options.Database != 0)
        {
            await client.ExecuteAsync(ct, "SELECT", options.Database.ToString()).ConfigureAwait(false);
        }

        return client;
    }

    /// <inheritdoc/>
    public async ValueTask DisposeAsync()
    {
        await _pool.DisposeAsync().ConfigureAwait(false);
    }

    // --- Internal command execution ---

    internal async Task<object?> ExecuteAsync(CancellationToken ct, params string[] args)
    {
        Exception? lastErr = null;
        for (var attempt = 0; attempt <= _options.MaxRetries; attempt++)
        {
            if (attempt > 0)
            {
                await Task.Delay(_options.RetryBackoff * attempt, ct).ConfigureAwait(false);
            }

            var conn = await _pool.GetAsync(ct).ConfigureAwait(false);
            try
            {
                await RespProtocol.WriteCommandAsync(conn.Stream, args, ct).ConfigureAwait(false);
                var result = await RespProtocol.ReadValueAsync(conn.Reader, ct).ConfigureAwait(false);
                _pool.Return(conn);
                return result;
            }
            catch (Exception ex)
            {
                lastErr = ex;
                conn.Dispose();
            }
        }
        throw lastErr!;
    }

    // --- String operations ---

    /// <summary>Gets the value of a key. Returns null if the key does not exist.</summary>
    public async Task<string?> GetAsync(string key, CancellationToken ct = default)
    {
        var val = await ExecuteAsync(ct, "GET", key).ConfigureAwait(false);
        return val?.ToString();
    }

    /// <summary>Sets a key to a string value.</summary>
    public async Task SetAsync(string key, string value, TimeSpan? expiry = null, CancellationToken ct = default)
    {
        if (expiry.HasValue)
        {
            await ExecuteAsync(ct, "SET", key, value, "PX",
                ((long)expiry.Value.TotalMilliseconds).ToString()).ConfigureAwait(false);
        }
        else
        {
            await ExecuteAsync(ct, "SET", key, value).ConfigureAwait(false);
        }
    }

    /// <summary>Deletes one or more keys. Returns the number removed.</summary>
    public async Task<long> DeleteAsync(CancellationToken ct = default, params string[] keys)
    {
        var args = new string[keys.Length + 1];
        args[0] = "DEL";
        Array.Copy(keys, 0, args, 1, keys.Length);
        return (long)(await ExecuteAsync(ct, args).ConfigureAwait(false))!;
    }

    /// <summary>Returns the number of specified keys that exist.</summary>
    public async Task<long> ExistsAsync(CancellationToken ct = default, params string[] keys)
    {
        var args = new string[keys.Length + 1];
        args[0] = "EXISTS";
        Array.Copy(keys, 0, args, 1, keys.Length);
        return (long)(await ExecuteAsync(ct, args).ConfigureAwait(false))!;
    }

    /// <summary>Sets a timeout on a key.</summary>
    public async Task<bool> ExpireAsync(string key, TimeSpan ttl, CancellationToken ct = default)
    {
        var val = await ExecuteAsync(ct, "EXPIRE", key,
            ((long)ttl.TotalSeconds).ToString()).ConfigureAwait(false);
        return (long)val! == 1;
    }

    /// <summary>Returns the remaining TTL of a key.</summary>
    public async Task<TimeSpan> TtlAsync(string key, CancellationToken ct = default)
    {
        var val = await ExecuteAsync(ct, "TTL", key).ConfigureAwait(false);
        return TimeSpan.FromSeconds((long)val!);
    }

    // --- List operations ---

    /// <summary>Prepends values to a list.</summary>
    public async Task<long> LPushAsync(string key, CancellationToken ct = default, params string[] values)
    {
        var args = new string[values.Length + 2];
        args[0] = "LPUSH"; args[1] = key;
        Array.Copy(values, 0, args, 2, values.Length);
        return (long)(await ExecuteAsync(ct, args).ConfigureAwait(false))!;
    }

    /// <summary>Appends values to a list.</summary>
    public async Task<long> RPushAsync(string key, CancellationToken ct = default, params string[] values)
    {
        var args = new string[values.Length + 2];
        args[0] = "RPUSH"; args[1] = key;
        Array.Copy(values, 0, args, 2, values.Length);
        return (long)(await ExecuteAsync(ct, args).ConfigureAwait(false))!;
    }

    /// <summary>Removes and returns the first element.</summary>
    public async Task<string?> LPopAsync(string key, CancellationToken ct = default)
    {
        var val = await ExecuteAsync(ct, "LPOP", key).ConfigureAwait(false);
        return val?.ToString();
    }

    /// <summary>Removes and returns the last element.</summary>
    public async Task<string?> RPopAsync(string key, CancellationToken ct = default)
    {
        var val = await ExecuteAsync(ct, "RPOP", key).ConfigureAwait(false);
        return val?.ToString();
    }

    /// <summary>Returns elements in the given range.</summary>
    public async Task<string[]> LRangeAsync(string key, long start, long stop, CancellationToken ct = default)
    {
        var val = await ExecuteAsync(ct, "LRANGE", key, start.ToString(), stop.ToString()).ConfigureAwait(false);
        return ToStringArray(val);
    }

    /// <summary>Returns the length of a list.</summary>
    public async Task<long> LLenAsync(string key, CancellationToken ct = default)
    {
        return (long)(await ExecuteAsync(ct, "LLEN", key).ConfigureAwait(false))!;
    }

    // --- Hash operations ---

    /// <summary>Sets field-value pairs in a hash.</summary>
    public async Task<long> HSetAsync(string key, CancellationToken ct = default, params string[] fieldValues)
    {
        var args = new string[fieldValues.Length + 2];
        args[0] = "HSET"; args[1] = key;
        Array.Copy(fieldValues, 0, args, 2, fieldValues.Length);
        return (long)(await ExecuteAsync(ct, args).ConfigureAwait(false))!;
    }

    /// <summary>Gets the value of a hash field.</summary>
    public async Task<string?> HGetAsync(string key, string field, CancellationToken ct = default)
    {
        var val = await ExecuteAsync(ct, "HGET", key, field).ConfigureAwait(false);
        return val?.ToString();
    }

    /// <summary>Returns all fields and values of a hash.</summary>
    public async Task<Dictionary<string, string>> HGetAllAsync(string key, CancellationToken ct = default)
    {
        var val = await ExecuteAsync(ct, "HGETALL", key).ConfigureAwait(false);
        return ToStringDictionary(val);
    }

    /// <summary>Deletes fields from a hash.</summary>
    public async Task<long> HDelAsync(string key, CancellationToken ct = default, params string[] fields)
    {
        var args = new string[fields.Length + 2];
        args[0] = "HDEL"; args[1] = key;
        Array.Copy(fields, 0, args, 2, fields.Length);
        return (long)(await ExecuteAsync(ct, args).ConfigureAwait(false))!;
    }

    // --- Set operations ---

    /// <summary>Adds members to a set.</summary>
    public async Task<long> SAddAsync(string key, CancellationToken ct = default, params string[] members)
    {
        var args = new string[members.Length + 2];
        args[0] = "SADD"; args[1] = key;
        Array.Copy(members, 0, args, 2, members.Length);
        return (long)(await ExecuteAsync(ct, args).ConfigureAwait(false))!;
    }

    /// <summary>Returns all members of a set.</summary>
    public async Task<string[]> SMembersAsync(string key, CancellationToken ct = default)
    {
        var val = await ExecuteAsync(ct, "SMEMBERS", key).ConfigureAwait(false);
        return ToStringArray(val);
    }

    /// <summary>Removes members from a set.</summary>
    public async Task<long> SRemAsync(string key, CancellationToken ct = default, params string[] members)
    {
        var args = new string[members.Length + 2];
        args[0] = "SREM"; args[1] = key;
        Array.Copy(members, 0, args, 2, members.Length);
        return (long)(await ExecuteAsync(ct, args).ConfigureAwait(false))!;
    }

    /// <summary>Returns the number of members in a set.</summary>
    public async Task<long> SCardAsync(string key, CancellationToken ct = default)
    {
        return (long)(await ExecuteAsync(ct, "SCARD", key).ConfigureAwait(false))!;
    }

    // --- Sorted Set operations ---

    /// <summary>Adds a member with a score.</summary>
    public async Task<long> ZAddAsync(string key, double score, string member, CancellationToken ct = default)
    {
        return (long)(await ExecuteAsync(ct, "ZADD", key,
            score.ToString(CultureInfo.InvariantCulture), member).ConfigureAwait(false))!;
    }

    /// <summary>Returns members in the given rank range.</summary>
    public async Task<string[]> ZRangeAsync(string key, long start, long stop, CancellationToken ct = default)
    {
        var val = await ExecuteAsync(ct, "ZRANGE", key, start.ToString(), stop.ToString()).ConfigureAwait(false);
        return ToStringArray(val);
    }

    /// <summary>Returns the score of a member.</summary>
    public async Task<double?> ZScoreAsync(string key, string member, CancellationToken ct = default)
    {
        var val = await ExecuteAsync(ct, "ZSCORE", key, member).ConfigureAwait(false);
        if (val == null) return null;
        return double.Parse(val.ToString()!, CultureInfo.InvariantCulture);
    }

    /// <summary>Removes members from a sorted set.</summary>
    public async Task<long> ZRemAsync(string key, CancellationToken ct = default, params string[] members)
    {
        var args = new string[members.Length + 2];
        args[0] = "ZREM"; args[1] = key;
        Array.Copy(members, 0, args, 2, members.Length);
        return (long)(await ExecuteAsync(ct, args).ConfigureAwait(false))!;
    }

    // --- Server operations ---

    /// <summary>Sends PING to verify connectivity.</summary>
    public async Task PingAsync(CancellationToken ct = default)
    {
        await ExecuteAsync(ct, "PING").ConfigureAwait(false);
    }

    /// <summary>Returns server information.</summary>
    public async Task<string> InfoAsync(CancellationToken ct = default)
    {
        var val = await ExecuteAsync(ct, "INFO").ConfigureAwait(false);
        return val?.ToString() ?? "";
    }

    // --- Helpers ---

    private static string[] ToStringArray(object? val)
    {
        if (val is not object?[] arr) return Array.Empty<string>();
        return arr.Select(o => o?.ToString() ?? "").ToArray();
    }

    private static Dictionary<string, string> ToStringDictionary(object? val)
    {
        if (val is not object?[] arr) return new Dictionary<string, string>();
        var dict = new Dictionary<string, string>();
        for (var i = 0; i < arr.Length - 1; i += 2)
        {
            var k = arr[i]?.ToString() ?? "";
            var v = arr[i + 1]?.ToString() ?? "";
            dict[k] = v;
        }
        return dict;
    }
}
